### ISSUE TITLE
T.proc.bind does not work if the block param is also T.nilable(...)

### ISSUE THREAD

**Issue by XrXr** [sorbet/sorbet#498]
### Problem

It is possible to rebind the type of `self` in a block: 

```ruby
# typed: true
class SelfSwitch
  extend T::Sig
  sig { params(block: T.proc.bind(Integer).void).void }
  def foo(&block)
  end
end


SelfSwitch.new.foo do
  T.reveal_type(self)   # Revealed type: Integer
end
```

So it seems natural to allow this capability for optional blocks too:

```ruby
# typed: true
class SelfSwitch
  extend T::Sig

  # Using `bind` is not permitted here https://sorbet.org/docs/error-reference#5004
  sig { params(block: T.nilable(T.proc.bind(Integer).void)).void }
  def optional_block(&block)
  end
end

```

**Comment by manhhung741** [comment]
@AlexEvanczuk @trevor-stripe  The issue with `T.nilable` is reported here. @XrXr did a good job filing the bug already. I also find it doesn't work with `T.any` (T.nilable is just T.any under the hood).

Here is a [sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aextend%20T%3A%3ASig%0Aclass%20A%0A%20%20def%20m%3B%20end%0Aend%0A%0Asig%20%7B%20params(block%3A%20T.nilable(T.proc.bind(A).void)).void%20%7D%0Adef%20run(%26block)%0A%20%20self.m%0Aend%0A%0Arun%20%7B%20'Hello'%20%7D)

**Comment by deecewan** [comment]
It also can't be used inside of `T.type_alias`

**Comment by connorshea** [comment]
This is making it hard to create a type definition for graphql-ruby :(

**Comment by deecewan** [comment]
@connorshea that's where I first ran into this issue as well.

**Comment by L-as** [comment]
Are there any major blockers for implementing this, or has it just gone without attention? Wondering if it's possible for me to fix it.

**Comment by deecewan** [comment]
@L-as I think it's just going to be difficult (based on the `hard` label).

**Comment by djrodgerspryor** [comment]
I'm having a crack at writing types for [Sequel](https://github.com/jeremyevans/sequel) and this has basically stopped me dead — Sequel makes heavy use of optional blocks with custom bindings.

Is there any workaround for this?

I'd be interested in making a PR to try and support this properly in Sorbet, but the 'hard' label makes me suspect that there are known problems — it would be super helpful if they were documented here @jez

**Comment by connorshea** [comment]
👀 is this fixed now because of #3859
